nn.functional.linear OpInfo

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64282 OpInfos for some Tensor dtype conversion methods
* #64214 Opinfos for avg_pooling
* **#61971 nn.functional.linear OpInfo**

Test Plan:
- wait for tests

Differential Revision: [D30013750](https://our.internmc.facebook.com/intern/diff/D30013750)

zou3519 committed:
nn.functional.linear OpInfo

Test Plan:
- wait for tests

[ghstack-poisoned]

zou3519 committed:
Update on "nn.functional.linear OpInfo"


Test Plan:
- wait for tests

[ghstack-poisoned]

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2154,6 +2154,32 @@ def sample_inputs_hardswish(self, device, dtype, requires_grad):
+
+    sample_inputs = []
+    for (in_feat, out_feat), batch_shape in itertools.product(features_options, batch_options):
+        input_tensor = make_tensor(batch_shape + [in_feat], device=device,
```
Reminder that as a style thing if this partials `make_tensor` it would be more succinct:

```
make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad, low=-2, high=2)

...

input_tensor = make_arg(batch_shape + [in_feat])
weight = make_arg([out_feat, in_feat])
...
```

zou3519 replied:
Thanks for the reminder! partial make it a lot more succinct

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -6444,6 +6470,22 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
            supports_gradgrad=True,
            supports_out=False,
            autodiff_nonfusible_nodes=["aten::leaky_relu"]),
+    OpInfo('nn.functional.linear',
```
I would add a comment before these entries for why they're separate. Typically we only separate an op into multiple entries if it has different metadata (like different dtypes) for different inputs. I imagine there's a reason this wants to treat linear with and without bias as separate operations?

zou3519 replied:
I don't have a good reason for why these should be separate, I was acting on reflex. I combined the two cases and now am modeling this as a single op; if someone else comes along with a use case to separate them in the future we could do that then.

mruberry commented:
Overall looks really good, @zou3519, but lint failures are real and looks like linear doesn't support bfloat16 on CUDA.

Not sure if this should be modeled as one op where some samples include bias and others don't or there's a reason to model this as two ops.

zou3519 committed:
Update on "nn.functional.linear OpInfo"


Test Plan:
- wait for tests

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30013750).

zou3519 committed:
Update on "nn.functional.linear OpInfo"


Test Plan:
- wait for tests

Differential Revision: [D30013750](https://our.internmc.facebook.com/intern/diff/D30013750)

[ghstack-poisoned]

zou3519 committed:
Update on "nn.functional.linear OpInfo"


Test Plan:
- wait for tests

Differential Revision: [D30013750](https://our.internmc.facebook.com/intern/diff/D30013750)

[ghstack-poisoned]

zou3519 committed:
Update on "nn.functional.linear OpInfo"


Test Plan:
- wait for tests

Differential Revision: [D30013750](https://our.internmc.facebook.com/intern/diff/D30013750)

[ghstack-poisoned]

zou3519 committed:
Update on "nn.functional.linear OpInfo"


Test Plan:
- wait for tests

Differential Revision: [D30013750](https://our.internmc.facebook.com/intern/diff/D30013750)

[ghstack-poisoned]

zou3519 committed:
Update on "nn.functional.linear OpInfo"


Test Plan:
- wait for tests

Differential Revision: [D30013750](https://our.internmc.facebook.com/intern/diff/D30013750)

[ghstack-poisoned]

zou3519 commented:
This is ready for review again (I finally figured out the bfloat16 failures... they are CUDA version dependent)

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30013750).

heitorschueroff reviewed (approved):
LGTM!

zou3519 committed:
Update on "nn.functional.linear OpInfo"


Test Plan:
- wait for tests

Differential Revision: [D30013750](https://our.internmc.facebook.com/intern/diff/D30013750)

[ghstack-poisoned]
